pkg/storer/internal/reserve/reserve.go

```diff
@@ -186,7 +208,16 @@ func (r *Reserve) Put(ctx context.Context, chunk swarm.Chunk) error {
 			return err
 		}
 
-		err = s.ChunkStore().Put(ctx, chunk)
+		has, err := s.ChunkStore().Has(ctx, chunk.Address())
```

Comment:
checking against the chunkstore wont work because the cache, pin, and the uploadstore also shares the chunkstore with the reserve, so you may end up with false positives. 

please do not deviate from the pseudo code here https://github.com/ethersphere/bee/compare/master...pseudocode-reserve-fix